File: 주다애/DP/prob5.java
class Solution {
    public int solution(int[] money) {
        int answer = 0;
        int len = money.length;
        if(len == 3) {
            return 0;
        }
        int[][] res = new int[2][len + 1];
        // 첯 번째 집 털음
        res[0][0] = money[0];
        // 첫 번째 집 안 털음
        res[1][0] = 0;
        res[1][1] = money[1];
        res[0][1] = money[0];
        
        for(int i = 2; i < len; i++) {
            res[0][i] = Math.max(money[i] + res[0][i - 2], res[0][i - 1]);
            res[1][i] = Math.max(money[i] + res[1][i - 2], res[1][i - 1]);
        }
        // 첫 번째 집과 마지막 집은 인접하므로 같이 못 털음
        return Math.max(res[0][len - 2], res[1][len - 1]);
    }
}

File: 전예진/DP/prob5.py
def solution(money):
    if len(money) == 3:
        return max(money)

    def dp(start, end):
        # 첫 집을 털었을 경우
        stole = [money[start], max(money[start], money[start + 1])]
        
        for i in range(start + 2, end):
            stole.append(max(stole[-1], stole[-2] + money[i]))
        
        return stole[-1]

    # 첫 집을 털었을 때와 안 털었을 때를 나눠서 최댓값 구하기
    return max(dp(0, len(money) - 1), dp(1, len(money)))
